File: examples/search-algorithms/sut-map-search.ts
/**
 * Map Search SUT
 *
 * Builds a Map from the data array (value to index) then performs
 * a single .get() lookup. Map construction time is included in
 * the execution time to reflect realistic single-lookup cost.
 */

interface SearchInput {
	data: number[];
	sortedData: number[];
	target: number;
}

interface SearchResult {
	found: boolean;
	index: number;
	comparisons: number;
	executionTimeMs: number;
}

export function createSut() {
	return {
		id: "map-search",
		config: {},
		run: (input: SearchInput): Promise<SearchResult> => {
			const { data, target } = input;
			const start = performance.now();

			const map = new Map<number, number>();
			for (let i = 0; i < data.length; i++) {
				map.set(data[i], i);
			}

			const result = map.get(target);
			const found = result !== undefined;
			const index = found ? result : -1;

			const executionTimeMs = performance.now() - start;

			return Promise.resolve({
				found,
				index,
				comparisons: 1,
				executionTimeMs,
			});
		},
	};
}

File: examples/search-algorithms/sut_map_search.py
"""Map Search SUT — dict lookup (construction time included)."""

import time


def create_sut():
    def run(input):
        data = input["data"]
        target = input["target"]
        start = time.perf_counter()

        lookup = {}
        for i in range(len(data)):
            lookup[data[i]] = i

        result = lookup.get(target)
        found = result is not None
        index = result if found else -1

        execution_time_ms = (time.perf_counter() - start) * 1000

        return {
            "found": found,
            "index": index,
            "comparisons": 1,
            "executionTimeMs": execution_time_ms,
        }

    return {"id": "map-search", "config": {}, "run": run}
